evaluation/utils.py

```diff
@@ -0,0 +1,135 @@
+        val_list = ps_dist[name]
+        assert val_list is not None and len(val_list) > 0, 'Parameter space should not be empty.\nCheck the configuration file.'
+        
+        xml_level = list(name.strip().split('.'))
+
+        # Linear scale range
+        if (val_list[0] == '.range'):
+            def convert_range(range_list):
+                '''Convert a list starting with .range to linear range.
+
+                Parameters:
+                ------------
+                range_list : List[Any]
+                The range statement. First should be .range; should be 3 or 4 in length.
+                
+                Returns:
+                -----------
+                values : List[Num]
+                The list of numeric values generated.
+                '''
+                LENGTH_NO_STEP = 3
+                LENGTH_WITH_STEP = 4
+                assert len(range_list) in (LENGTH_NO_STEP, LENGTH_WITH_STEP), 'Incorrect argument number for linear scale.\nCheck the configuration file.'
+                if (len(range_list) == LENGTH_NO_STEP):
+                    start, end, step = range_list[1], range_list[2], 1
+                else:
+                    start, end, step = range_list[1:]
+
+                values = np.arange(start, end, step).tolist()
+                return values
+
+            val_list = convert_range(val_list)
+
+        ps_space.append((xml_level, val_list))
+
+    return ps_space
+
+def inject_param_xml(file_path, parameters):
+    '''Inject and re-write XML file with given parameters.
+
+    Parameters:
+    -----------
+    file_path : str
+        XML file path to inject.
+    parameters : List[Tuple(List[str], Any)]
+        The list of parameter names and values to inject.
+    '''
+    conf_etree = ET.parse(file_path)
+    root = conf_etree.getroot()
+    for name_level, val in parameters:
+        cursor = root
+        # Traverse XML name levels.
+        for key in name_level:
+            cursor = cursor.find(key)
+            if cursor is None:
+                break
+
+        assert cursor is not None, 'Fail to inject parameter in conf file,' + str(name_level) + '\nCheck the format of target XML file.'
+        cursor.text = str(val)
+
+    conf_etree.write(file_path)
+
+def parameter_sweep(ps_space, f, closure=None):
+    '''Recursive calling routine on parameter space.
+
+    Parameters:
+    ------------
+    ps_space : List[Tuple(List[str], List[Any])]
+        Parameter space to sweep. each element is parameter name level + value list.
+    f : (List[Tuple(List[str], Any)], Dict[str, Any])->Any
+        Callback function to be executed in the sweep, takes parameter combination
+        and closure dict.
+    closure : Dict[str, Any]
+        Closure environment passed from caller.
+    '''
+    assert(len(ps_space) > 0), 'Parameter space should not be empty.\nCheck the configuration file.'
+    # Maintain the traverse states. Initial to the first value on level 0.
```

Comment:
```suggestion

    if closure is None:
        closure = {}

    # Maintain the traverse states. Initial to the first value on level 0.
```

I think you missed this part